documentation/writing_tests.md

```diff
@@ -0,0 +1,1473 @@
+```js
+  browserConsoleLogOptions: {
+    terminal: !process.env.TEST_MODE
+  },
+```
+
+You can see an example [karma.config.js in our template project](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/template/template/karma.conf.js).
+
+### The Recorder
+
+The Azure SDK for JavaScript and TypeScript uses a custom utility to record tests that hit the live endpoints, so that these tests can be executed almost instantly against these recordings instead of hitting the live services. This tool is called [@azure/test-utils-recorder](https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/test-utils/recorder). It's an unpublished package that can be added using rush, as follows:
+
+```
+rush add --dev -p @azure/test-utils-recorder
+```
+
+You can read more about the recorder in its readme: https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/test-utils/recorder
+
+
+
+## Package configuration
+
+Tests for the Azure SDK for JavaScript and TypeScript should be all executed using the NPM command `test`, they should be stored in the `test` folder, they must be written in TypeScript, in files that will end in `.spec.ts`. Our [recommended tools](#recommended-tools) should provide the test framework and proper ways of compiling and bundling our tests correctly. They should be able to run and be debugged on the environments we're targeting, which are currently only NodeJS and the browsers.
+
+Even though tests can be executed in groups, through the use of [Pattern matching test titles](#pattern-matching-test-titles), tests should be considered to be executed all at once. What determines what will be tested will be how tests are written, either through [the recorder](#the-recorder) or through [using conditionals](#using-conditionals). The execution of tests will vary in the `package.json` to provide different build environments, different debugging mechanisms and reportings for automated live tests or automated verification tests. To read more about these topics, you can go to the sections [Configuring Mocha](#configuring-mocha), [Configuring Karma](#configuring-karma) and [CI and nightly test configuration](#ci-and-nightly-test-configuration).
+
+All together, the `package.json` will end up with the following scripts:
+
+```json
+  "test": "npm run build:test && npm run test:node && npm run test:browser",
+  "test-ci": "npm run build:test && npm run test-ci:node && npm run test-ci:browser",
+
+  "test:node": "mocha --require source-map-support/register --reporter ../../../common/tools/mocha-multi-reporter.js --timeout 180000 --full-trace dist-test/index.node.js",
+  "test-ci:node": "nyc mocha -r esm --require source-map-support/register --reporter ../../../common/tools/mocha-multi-reporter.js --timeout 180000 --full-trace \"dist-esm/test/**/*.spec.js\"",
```

Comment:
- Are we enforcing this for all the packages?
- So far, we agreed upon `test` command being the convenience for the developers by making `clean`, `build:test` part of it. Are we taking that away?

Reply:
I will add `npm run clean` too! This is a proposal after all, so I am proposing to have all of what I'm putting here, yes. If that helps.

Reply:
Here's how these two are looking with your feedback:

```
  "test": "npm run clean && npm run build:test && npm run test:node && npm run test:browser",
  "test-ci": "npm run clean && npm run build:test && npm run test-ci:node && npm run test-ci:browser",
```

Reply:
This is still taking away the convenience of not allowing executing only the node tests with "clean" and "build:test".

Reply:
Thanks! I've moved them to this:

```json
  "test": "npm run test:node && npm run test:browser",
  "test-ci": "npm run test-ci:node && npm run test-ci:browser",

  // Where the individual commands are:
  "test:node": "npm run clean && npm run build:test && mocha --require source-map-support/register --reporter ../../../common/tools/mocha-multi-reporter.js --timeout 180000 --full-trace dist-test/index.node.js",
  "test-ci:node": "npm run clean && npm run build:test && nyc mocha -r esm --require source-map-support/register --reporter ../../../common/tools/mocha-multi-reporter.js --timeout 180000 --full-trace dist-esm/test/**/*.spec.js",
  // Remember to specify different karma.config.js if needed
  "test:browser": "npm run clean && npm run build:test && karma start --single-run",
  "test-ci:browser": "npm run clean && npm run build:test && npm run test:browser",

  // Where we're cleaning with:
  "clean": "rimraf dist dist-* types",
```

Reply:
Resolving this in favor of the other thread.